state/sqlserver/migration.go

```diff
@@ -275,32 +274,61 @@ func (m *migration) ensureUpsertStoredProcedureExists(db *sql.DB, mr migrationRe
 		CREATE PROCEDURE %s (
 			@Key 			%s,
 			@Data 			NVARCHAR(MAX),
-			@RowVersion 	BINARY(8))
+			@RowVersion 	BINARY(8),
+			@FirstWrite		BIT)
 		AS
 			IF (@RowVersion IS NOT NULL)
```

Comment:
This initial update if any rowversion is provided seems wrong. It basically short circuits everything else to try an update (which I believe can fail if the row doesn't already exist) requiring that the rowversion matches. Looking at the overall logical flow of the process, it seems like it should:

- Try to insert without using the etag
	- If it succeeds, key did not previously exist, and we correctly ignored an irrelevant etag if provided 
		- In the future, this should fail if not also first-write, but that is a breaking change incoming after https://github.com/dapr/dapr/issues/2739, option 1c.
	- Else, key exists and operation should be an update.
		- If first-write and Etag does not match, fail the operation
		- Else update the row with the optional Etag (via the ISNULL construction)

Reply:
What if FirstWrite=1 and RowVersion is NULL? It feels that FirstWrite check has higher priority over RowVersion.

Reply:
@CodeMonkeyLeet I cannot manually set an Etag. The Etag is added by the stored procedure only. Any attempt to manually set an etag miserable failed in that the upsert was completely ignored -- I spent one full day just on that.

Does that change what you are suggesting here?

Reply:
Interpret ETag as RowVersion in my comment, it's just an input into the method. I'm not suggesting adding an ETag where one was not provided in the request.

Reply:
I agree with you both now @CodeMonkeyLeet @artursouza. Sorry for my confusion.

**I incorrectly thought the First Write concept only applied to database records that do not exist in the database.**